app/src/main/java/org/cuberite/android/MainActivity.java

```diff
@@ -107,35 +107,33 @@ public static void showSnackBar(Activity activity, String message) {
                 .show();
     }
 
-    private void showPermissionPopup() {
-        AlertDialog.Builder permissionPopupBuilder = new AlertDialog.Builder(this);
-        permissionPopupBuilder.setTitle(getString(R.string.status_permissions_needed));
-        permissionPopupBuilder.setMessage(R.string.message_externalstorage_permission);
-        permissionPopupBuilder.setCancelable(false);
-        permissionPopupBuilder.setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
-            public void onClick(DialogInterface dialog, int id) {
-                Log.d(LOG, "Requesting permissions for external storage");
-                ActivityCompat.requestPermissions(MainActivity.this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, REQUEST_WRITE_EXTERNAL_STORAGE_PERMISSION);
-            }
-        });
+   private void showPermissionPopup() {
+        permissionPopup = new AlertDialog.Builder(this)
+            .setTitle(getString(R.string.status_permissions_needed))
+            .setMessage(R.string.message_externalstorage_permission)
+            .setCancelable(false)
+            .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
+                public void onClick(DialogInterface dialog, int id) {
+                    Log.d(LOG, "Requesting permissions for external storage");
+                    ActivityCompat.requestPermissions(MainActivity.this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, REQUEST_WRITE_EXTERNAL_STORAGE_PERMISSION);
+                }
+            })
+            .create();
 
-        permissionPopup = permissionPopupBuilder.create();
         permissionPopup.show();
     }
 
     private void checkPermissions() {
-        if (preferences.getString("cuberiteLocation", null) == null) {
-            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
-                // Always use public dir in Lollipop and earlier, since permissions are granted when the app is installed
-                preferences.edit().putString("cuberiteLocation", PUBLIC_DIR + "/cuberite-server").apply();
-            } else {
+        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {

```

Comment:
I do not like the reverse order of the `if` statements here, on Android <= 6 they lead to an unnecessary file write on every single `checkPermissions()` call, which is quite sprawled all over the place.

Two solutions would be either reverting to the not-null check before everything else or alternatively check in the else branch below ([L 137](https://github.com/cuberite/android/pull/39/files#diff-53dcaef8d195b3ca85811c12b44f377eR137)) whether the preferences show the correct location before writing to the preferences file.

Reply:
Good point! I don't remember how it used to be, but checkPermissions is only called when the app resumes.